### PR TITLE
enable race detector by default in make test

### PR DESCRIPTION
**Pull request by BenTheElder**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

enables the golang race detector by default when running unit tests
see: https://github.com/kubernetes/kubernetes/issues/102607

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

with this change we'll be able to change CI to actually run `make test` with no parameters, bringing it closer to trivially reproducing CI locally ™️ 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```
/sig testing
/triage accepted
/priority important-longterm

### CODE REVIEW

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/hold
remove hold when ready to merge (I dunno if you wanted to run the presubmit a few times to see if it starts flaking a bunch)

**Comment by BenTheElder** [comment]
> remove hold when ready to merge (I dunno if you wanted to run the presubmit a few times to see if it starts flaking a bunch)

actually in CI we run with `make test KUBE_RACE=-race` already, this will allow us to switch CI to `make test` in a follow-up (and ensure that in the future CI just runs what you'd run locally, this is the last thing we tweak in CI vs local defaults)

https://github.com/kubernetes/test-infra/blob/a5528997fea7e64a44a3335ad95c4d3df6514ed7/config/jobs/kubernetes/sig-testing/make-test.yaml#L24

/hold cancel

**Comment by BenTheElder** [comment]
very confused why `pull-kubernetes-integration` is failing

we already set `export KUBE_RACE=-race` in `hack/jenkins/test-dockerized.sh`, which is what we run in the job config.

**Comment by BenTheElder** [comment]
this gem courtesy @liggitt 
https://github.com/kubernetes/kubernetes/blob/f1d9db4381a7dd383894e32193bccd8d994f7023/hack/make-rules/test-integration.sh#L69-L78

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/issues/99877 https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102960/pull-kubernetes-unit/1405766620067926016
/retest

**Comment by BenTheElder** [comment]
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/102960/pull-kubernetes-integration/1405779343971454976/build-log.txt

> +++ [0618 06:55:34] Running tests without code coverage and with -race

....

**Comment by BenTheElder** [comment]
I suspect make does not pass an empty env as set to empty, triggering the default path.

Will revisit in the morning. one possibility is to move defaulting into the makefile as a make variable

File: hack/make-rules/test.sh
Comment (BenTheElder) [inline hack/make-rules/test.sh:76]: we usually use `${FOO:-default}` but this will default in the set-but-null case. We need `${FOO-default}` here (which will not override set-but-empty) so you can get the previous behavior with `KUBE_RACE=""` from the commandline/env/...

**Comment by BenTheElder** [comment]
/retest
node-e2e flake https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102960/pull-kubernetes-node-e2e-containerd/1406006552288038912/

**Comment by BenTheElder** [comment]
unit and integration happy. you can see that race is on/off as expected in the logs which now log when -race is enabled while testing in the message about coverage.

File: hack/jenkins/test-dockerized.sh
Comment (BenTheElder) [inline hack/jenkins/test-dockerized.sh:22]: nothing calls this for O(years), dead code

File: hack/jenkins/test-dockerized.sh
Comment (BenTheElder) [inline hack/jenkins/test-dockerized.sh:51]: we set this elsewhere. this is redundant / useless / another way to make this more confusing.

File: hack/jenkins/test-dockerized.sh
Comment (BenTheElder) [inline hack/jenkins/test-dockerized.sh:57]: the scripts all default this to match artifacts unless otherwise set. simplify this script a bit more.

ideally test-dockerized should go away, but integration job still kinda needs it.

File: hack/make-rules/test.sh
Comment (amwat) [reply hack/make-rules/test.sh:76]: make this the code comment?
or add a link to https://wiki.bash-hackers.org/syntax/pe#use_a_default_value?

**Review by amwat — COMMENTED** [review]
/lgtm

File: hack/make-rules/test.sh
Comment (BenTheElder) [reply hack/make-rules/test.sh:76]: I tried to not make this overly verbose, hopefully future editor will think twice before changing, we have history to point back here / reviewers should check that the result is sane.

it will also break CI if done wrong (which is why this PR took so long)

File: hack/make-rules/test-cmd.sh
Comment (liggitt) [inline hack/make-rules/test-cmd.sh:173]: gah....

File: hack/make-rules/test-integration.sh
Comment (liggitt) [inline hack/make-rules/test-integration.sh:74]: does this order change matter? I really hope it doesn't.

Comment (BenTheElder) [reply hack/make-rules/test-integration.sh:74]: no, just normalizing a bit with the ~matching invocation in test-cmd.sh, these went through a fair few iterations pinning down what was actually happening with these throughout the process tree.

**Pull request merged**
